### ISSUE TITLE
Consider adding a Preview flag that does not use secrets

### ISSUE THREAD

**Issue by AaronFriel** [pulumi/pulumi#12289]
As a maintainer of open source repositories, I want to ensure that a malicious pull request cannot subvert security controls or leak important secrets in my organization. GitHub has coined these as [pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/), and they are an ever-present threat when some code in a continuous integration/deployment tool may be untrusted.

In order to harden Pulumi Deployments, GitHub Actions, and other tools that rely on events from repositories, a "fully offline" Preview mode that does spin up provider binaries *and* does not require decrypting secrets would be extraordinarily helpful. In this mode, any secret value should be considered unknown, and for the purpose of calculating the diff any secrets in the initial state would be unknown as well.

This would enable running previews without any secrets, as GitHub Actions implemented for `pull_request` workflows.

In this mode we should consider two options:
* (Ideal) Whether it's possible to run preview without launching the provider plugin entirely.
* (Less ideal) RPC changes to signal to a provider that it should operate in a mode that may not have credentials.

The first is ideal because it does not require providers to implement any changes, centralizing this security property reduces the work on all future authors of providers. As shown below with the current behavior of `preview` using the AWS provider, providers are not currently authored in a way to signal this desired property via a single flag, and we may not be able to rely on provider authors to implement this offline mode correctly.

## Desired behavior

```sh
> pulumi new aws-go
# without any ambient credentials, and secrets in state sent to the program as unknowns:
> pulumi preview
Previewing update (dev)

View Live: https://app.pulumi.com/friel/tmp.r6yznvhqi0/dev/previews/6cfbfe65-0661-4012-9dea-6bfb59509a9a

     Type                 Name                Plan       
 +   pulumi:pulumi:Stack  tmp.r6yznvhqi0-dev  create     
 +   └─ aws:s3:Bucket     my-bucket           create     


Outputs:
    bucketName: output<string>

Resources:
    + 2 to create

```

## Current behavior

Unfortunately in our current Preview, providers may check for credentials even in a dry-run. E.g.: to run the AWS provider in a dry run, one has to set a combination of config values and set an environment variable:

```sh
> pulumi new aws-go
# without any ambient credentials:
> pulumi preview
# ...
Diagnostics:
  aws:s3:Bucket (my-bucket):
    error: unable to validate AWS credentials.
    Details: no valid credential sources for Pulumi AWS Classic found.
    
    Please see https://www.pulumi.com/registry/packages/aws/installation-configuration/
    for more information about providing credentials.
    
    Error: failed to refresh cached credentials, the SSO session has expired or is invalid
    
    Make sure you have set your AWS region, e.g. `pulumi config set aws:region us-west-2`.
```

It is possible to force the AWS provider into an offline mode for previews, however each provider has its own unique method of signaling this intent. For the AWS provider, one must set five config keys:

```sh
pulumi config set aws:skipCredentialsValidation true
pulumi config set aws:skipMetadataApiCheck      true
pulumi config set aws:skipRequestingAccountId   true
# These two may be any non-empty string:
pulumi config set --secret aws:accessKey "a non-empty value" 
pulumi config set --secret aws:accessKey "another non-empty value"
```

And then additionally use an environment variable to disable a fourth check in the lifecycle methods when running preview:

```sh
AWS_EC2_METADATA_DISABLED=true pulumi preview
```

This succeeds with:

```
Previewing update (dev)

View Live: https://app.pulumi.com/friel/tmp.r6yznvhqi0/dev/previews/6cfbfe65-0661-4012-9dea-6bfb59509a9a

     Type                 Name                Plan       
 +   pulumi:pulumi:Stack  tmp.r6yznvhqi0-dev  create     
 +   └─ aws:s3:Bucket     my-bucket           create     


Outputs:
    bucketName: output<string>

Resources:
    + 2 to create
```

**Comment by Frassle** [comment]
> (Ideal) Whether it's possible to run preview without launching the provider plugin entirely.

In general yes, we already skip calling into the provider if any of it's configuration options are unknown. But it leads to a really reduced preview experience. There's a lot the engine just can't get right about diffs without the provider telling it.

**Comment by AaronFriel** [comment]
Agreed! But I think it would be a valuable feature to opt in. If we already support something pretty close to that, it might be able to help us close another issue in Deploy.
